dts: bindings: mmu_mpu: Remove arm,num-mpu-regions

ithinuel opened the pull request:
The C code does not required this parameter to be defined and falls back to the MPU_TYPE's register to return the appropriate value.

This change enables the tail gating of a fix for the board mps2_an521.

The SSE-200 implementation on AN521 only has 8 regions. Although the AN521 FPGA image implements the documented value, qemu-system-arm currently shows in MPU_TYPE the incorrect value of 16.

Removing the property definition allows the code to rely on the register read value and work on both qemu-system-arm & the physical board.

ithinuel committed:
dts: bindings: mmu_mpu: Mark arm,num-mpu-regions as optional

The C code does not required this parameter to be defined and falls back
to the MPU_TYPE's register to return the appropriate value.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

ithinuel committed:
boards: arm: mps2_an521: remove the arm,num-mpu-regions property

The SSE-200 implementation on AN521 only has 8 regions.
Although the AN521 FPGA image implements the documented value,
qemu-system-arm currently shows in MPU_TYPE the incorrect value of 16.

Removing the property definition allows the code to rely on the register
read value and work on both qemu-system-arm & the physical board.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

gmarull reviewed (commented):
question: is the property used at all then?

carlocaione commented:
> question: is the property used at all then?

yeah, same question. It looks like it is only used for an assert in `arch/arm/core/aarch32/mpu/arm_mpu.c`?

ithinuel commented:
Indeed, As far as I can tell it is only used in that assert and in the implementation of `get_num_regions` in [‎arch/arm/core/aarch32/mpu/cortex_m/arm_mpu_internal.h](https://github.com/zephyrproject-rtos/zephyr/blob/4d56a318f162b758daf04525143380da373e1814/arch/arm/core/aarch32/mpu/cortex_m/arm_mpu_internal.h#L11) and [‎arch/arm/core/aarch32/mpu/cortex_a_r/arm_mpu_internal.h](https://github.com/zephyrproject-rtos/zephyr/blob/4d56a318f162b758daf04525143380da373e1814/arch/arm/core/aarch32/mpu/cortex_a_r/arm_mpu_internal.h#L11).

It may also be used in downstream application/libraries.

The property was flagged as required since the inception of the dts binding in [this commit](https://github.com/zephyrproject-rtos/zephyr/commit/0a08c4e1e9667a24b528c858ecbebe30ee933e8a).

ithinuel commented:
Also, I fail to identify the relation between the failures and the changes in this PR.

carlocaione reviewed (dismissed):
alright, let's make this optional.

pm215 commented:
The other option that you might consider is whether to make the assert() just assert "the core we're running on has at least this many regions" rather than "exactly this many" -- AFAIK a guest binary built for 8 MPU regions ought to run fine on an implementation with 16. This might be the way to go if it turns out there are downstream libraries/apps that depend on the #define.

Some side notes from the peanut gallery :-) :

(1) does get_num_region() really need to have special cases for M0+/M3/M4 that hardcode "8 regions"? I thought all those cores had an MPU_TYPE register (and indeed the 'sanity checks' in arm_mpu.c check that the MPU_TYPE register reads as 8 on those cores)

(2) if we can make the dtb property optional, should we remove it globally (i.e. for all boards)? That would be the fastest way to find out if anybody's depending on it being present...

ithinuel commented:
1. I'd assume that relying on a const rather than a register may allow for some optimisation from the compiler. Although, I wouldn't expect to gain much from this niche.
2. This could be an option but considering the number of target and unless all tests run on boards making sure the dts matches the hardware for all, I cannot afford to check all of them if the value in the dts is an intended deviation. Making the property optional at least makes it possible for future platform to omit it and existing ones to gradually be updated.

pm215 commented:
If the DTS doesn't have the hardware then AIUI at the moment Zephyr won't run at all -- that's how we found this problem...

ithinuel commented:
I discovered that there is a mismatch between the HW of AN521 and qemu's implementation (reported [here](https://gitlab.com/qemu-project/qemu/-/issues/1772)) which is a v8-m device.

On CM0+, CM3 and CM4, the assert checks the MPU_TYPE reports 8. So the dts may have another value, AFAIK this is not checked.

pm215 commented:
For M0+, M3 and M4, currently the code (a) always uses 8 and (b) checks that matches the MPU_TYPE register. So on those boards switching to a design of "always use MPU_TYPE register value, don't hardcode, don't look at num-mpu-regions" should be no-behaviour-change.

For other CPUs, if the dts doesn't specify the num-mpu-regions then we already use the MPU_TYPE register value, so that would also be no-behaviour-change. If the dts *does* specify num-mpu-regions, then currently we check that against MPU_TYPE and assert if it's different, so there shouldn't be any boards where only looking at MPU_TYPE would give us a different number from the value in the dts. 

The exception in both cases would be if there are any platforms where people are only building and running with assertions disabled and nobody's ever tried to run a Zephyr with assertions enabled on them.

The tricky question is whether anybody's using the NUM_MPU_REGIONS define outside Zephyr proper. If they are, then I'm not sure you can make num-mpu-regions optional.

ithinuel committed:
dts: arm: Blanket remove all usages of arm,num-mpu-regions

This removes all occurrences of arm,num-mpu-regions relying on the value
reported by the register instead.
A user may still define this property if they need to have a compile time
definition for it.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

ithinuel commented:
Just checking, do I need to edit anything on this PR?

povergoing reviewed (dismissed):
Making it optional looks good to me

galak reviewed (changes requested):
Changing to optional is fine for the dt prop.  I'm confused by the removal of the property in the dts files.  Are we saying the cases that set the value to 16 are wrong?

ithinuel commented:
They are not necessarily wrong but most likely overspecified.

On V8-M, the value is asserted against what the MPU_TYPE register returns.

pm215 commented:
> Changing to optional is fine for the dt prop. I'm confused by the removal of the property in the dts files. Are we saying the cases that set the value to 16 are wrong?

Setting the property in the dts files effectively does two things:
 * creates a #define (which we are assuming nobody downstream is using)
 * asserts at runtime that the hardware has the value that the dts says it does

Removing the property from the dts files means that Zephyr just asks the hardware (via the MPU_TYPE register) what its config is, rather than crashing at startup if it's different from the dts value, even though it could run just fine. Doing this for every dts file is IMHO the easiest way to find out whether anybody downstream really is using that #define...

(There's some followup cleanup where Zephyr could ignore the dts property entirely, rather than continuing with its current behaviour if the dts does provide it.)

galak reviewed (changes requested):
Based on the code, why don't we remove this property all together (and remove the assert check).

ithinuel committed:
arch: arm: remove num-mpu-regions

Remove this property presumably unused.

Signed-off-by: Wilfried Chauveau <wilfried.chauveau@arm.com>

gmarull commented on dts/bindings/mmu_mpu/arm,armv7m-mpu.yaml:
```diff
@@ -7,8 +7,3 @@ include: base.yaml
   reg:
     required: true
-
-  arm,num-mpu-regions:
-    required: true
-    type: int
-    description: number of MPU regions supported by hardware
```
considering this has been a strict requirement for any ARM SoC, maybe deprecation first would be better. I personally don't care, so will leave the decision to DT maintainers.

ithinuel replied:
removal of the property was suggested by @galak

povergoing commented:
I might ask a silly question, but I wonder if there is a use case in which we have to tell Zephyr to use limited regions (say up to 6 regions) although the PE has 8 regions, meaning that, in this case, we should use the value from DTS instead of MPU_TYPE?

ithinuel commented:
I cannot think of any right now, but either way, this property is documented as `number of MPU regions supported by hardware`.
IMHO, if the need for a `maximum allowed region` property araises (as device tree or Kconfig), this can still be introduced at a later stage.

povergoing commented:
> IMHO, if the need for a maximum allowed region property araises (as device tree or Kconfig), this can still be introduced at a later stage.

Right, this can be introduced if necessary.

ithinuel commented:
Thanky you all for your time and reviews :)
